File: Leetcode-Problems/Top K Frequent Elements.cpp
class Solution
{
public:
    vector<int> topKFrequent(vector<int> &nums, int k)
    {
        unordered_map<int, int> mp;
        for (int i : nums)
            mp[i]++;
        vector<pair<int, int>> arr(mp.begin(), mp.end());
        sort(arr.begin(), arr.end(), [](auto &a, auto &b)
             { return a.second > b.second; });
        vector<int> ans;
        for (int i = 0; i < k; i++)
        {
            ans.push_back(arr[i].first);
        }
        return ans;
    }
};

File: Leetcode-Problems/Top K Frequent Elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> mp = new HashMap<>();
        for (int i : nums) {
            mp.put(i, mp.getOrDefault(i, 0) + 1);
        }
        List<Map.Entry<Integer, Integer>> list = new ArrayList<>(mp.entrySet());
        Collections.sort(list, (a, b) -> b.getValue() - a.getValue());
        int[] ans = new int[k];
        for (int i = 0; i < k; i++) {
            ans[i] = list.get(i).getKey();
        }
        return ans;
    }
}

class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        HashMap<Integer, Integer> mp = new HashMap<>();
        for (int i : nums) {
            mp.put(i, mp.getOrDefault(i, 0) + 1);
        }

        PriorityQueue<Map.Entry<Integer, Integer>> pq = new PriorityQueue<>((a, b) -> a.getValue() - b.getValue());
        for (Map.Entry<Integer, Integer> entry : mp.entrySet()) {
            pq.add(entry);
            if (pq.size() > k) {
                pq.poll();
            }
        }

        int[] ans = new int[k];
        int i = 0;
        for (Map.Entry<Integer, Integer> entry : pq) {
            ans[i++] = entry.getKey();
        }
        return ans;
    }
}